dhis-2/dhis-web/dhis-web-api/src/main/java/org/hisp/dhis/webapi/controller/tracker/export/TrackerRelationshipsExportController.java

```diff
@@ -0,0 +1,166 @@
+
+    private final TrackedEntityInstanceService trackedEntityInstanceService;
+
+    private final ProgramInstanceService programInstanceService;
+
+    private final ProgramStageInstanceService programStageInstanceService;
+
+    private final RelationshipService relationshipService;
+
+    private static final RelationshipMapper RELATIONSHIP_MAPPER = Mappers.getMapper( RelationshipMapper.class );
+
+    private Map<Class<?>, Function<String, ?>> objectRetrievers;
+
+    private Map<Class<?>, Function<Object, List<Relationship>>> relationshipRetrievers;
+
+    @PostConstruct
+    void setupMaps()
+    {
+        objectRetrievers = ImmutableMap.<Class<?>, Function<String, ?>> builder()
+            .put( TrackedEntityInstance.class, trackedEntityInstanceService::getTrackedEntityInstance )
+            .put( ProgramInstance.class, programInstanceService::getProgramInstance )
+            .put( ProgramStageInstanceService.class, programStageInstanceService::getProgramStageInstance )
+            .build();
+
+        relationshipRetrievers = ImmutableMap.<Class<?>, Function<Object, List<Relationship>>> builder()
+            .put( TrackedEntityInstance.class,
+                o -> relationshipService.getRelationshipsByTrackedEntityInstance( (TrackedEntityInstance) o, false ) )
+            .put( ProgramStage.class,
+                o -> relationshipService.getRelationshipsByProgramInstance( (ProgramInstance) o, false ) )
+            .put( ProgramStageInstance.class,
+                o -> relationshipService.getRelationshipsByProgramStageInstance( (ProgramStageInstance) o, false ) )
+            .build();
+    }
+
+    @GetMapping( produces = APPLICATION_JSON_VALUE )
+    List<org.hisp.dhis.tracker.domain.Relationship> getInstances(
+        @RequestParam( required = false ) String tei,
+        @RequestParam( required = false ) String enrollment,
+        @RequestParam( required = false ) String event )
+        throws WebMessageException
+    {
+        List<org.hisp.dhis.tracker.domain.Relationship> relationships = tryGetRelationshipFrom( tei,
+            TrackedEntityInstance.class, () -> notFound( "No trackedEntityInstance '" + tei + "' found." ) )
+                .orElseGet( () -> tryGetRelationshipFrom( enrollment, ProgramInstance.class,
+                    () -> notFound( "No enrollment '" + enrollment + "' found." ) )
+                        .orElseGet( () -> tryGetRelationshipFrom( event, ProgramStageInstance.class,
+                            () -> notFound( "No event '" + event + "' found." ) )
+                                .orElse( null ) ) );
+
+        if ( Objects.isNull( relationships ) )
+        {
+            throw new WebMessageException( badRequest( "Missing required parameter 'tei', 'enrollment' or 'event'." ) );
+        }
+        return relationships;
+    }
+
+    @SneakyThrows
+    private Optional<List<org.hisp.dhis.tracker.domain.Relationship>> tryGetRelationshipFrom(
```

Comment:
Here the use of Optional seems redundant.
You could just use a list and have some actions when the list is empty.